File: CodeForces/271D/main.cpp
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <cstdlib>
#include <cstring>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <iomanip>
#include <queue>
#include <locale>
#include <deque>
#include <cmath>
using namespace std;

#define INF (1<<30)
#define SWAP(x, y) x = x ^ y; y = x ^ y; x = x ^ y;
#define MAX(a, b) ((a > b) ? (a) : (b))

struct vertex
{
    int     next[26];
    int     p;
    int     go[26];
    char    pch;
    bool    leaf;
    int     cnt;
    int     up;
    int     link;
    int     bc = 0;
};

inline void init(int k);
inline void add_string(const string& s);
inline int link(int v);
inline int go(int v, char c);
inline int cnt(int v);

vertex t[2000100];
int k = 1, ans = 0, bc;
bool good[26];


inline void init(int k)
{
    t[k].leaf   = false;
    t[k].p      = -1;
    t[k].cnt    = -1;
    t[k].link   = -1;
    for(int i = 0; i < 26; i++)
        t[k].next[i] = -1, t[k].go[i] = -1;
}
inline void add_string(const string &s)
{
    int v = 0;
    for(int i = 0; i < s.length(); i++)
    {
        char c = (int) s[i] - 'a';
        if(t[v].next[c] == -1)
        {
            init(k);
            t[k].p          = v;
            t[k].pch        = c;
            t[k].bc         = t[v].bc + ((good[c]) ? 0 : 1);
            t[v].next[c]    = k;
            if(t[k].bc <= bc)
                ans++;
            k++;
        }
        v = t[v].next[c];
    }
    t[v].leaf = true;
}
inline int link(int v)
{
    if(t[v].link == -1)
    {
        if(v == 0 || t[v].p == 0)
            t[v].link = 0;
        else
            t[v].link = go(link(t[v].p), t[v].pch);
    }
    return t[v].link;
}
inline int go(int v, char c)
{
    if(t[v].go[c] == -1)
    {
        if(t[v].next[c] != -1)
            t[v].go[c] = t[v].next[c];
        else if(v == 0)
            t[v].go[c] = 0;
        else
            t[v].go[c] = go(link(v), c);
    }
    return t[v].go[c];
}
inline int cnt(int v)
{
    if(t[v].cnt == -1)
    {
        if(t[v].cnt == 0)
            t[v].cnt = 0;
        else
            t[v].cnt = ((t[v].leaf) ? 1 : 0) + cnt(link(v));
    }
    return 0;
}


int main() {
//    freopen("/home/jyree/src/input", "r", stdin);
//    freopen("/home/jyree/src/output.txt", "w", stdout);
//    freopen("input.txt", "r", stdin);
//    freopen("output.txt", "w", stdout);
    string s1, s2;
    init(0);
    cin >> s1;
    cin >> s2;
    cin >> bc;
    for(int i = 0; i < 26; i++)
        good[i] = (s2[i] - '0');
    for(int i = 0; i < s1.length(); i++)
        add_string(s1.substr(i));
    cout << ans << endl;
    return 0;
}

File: CodeForces/271D/Main.java
import java.awt.image.AreaAveragingScaleFilter;
import java.io.PrintWriter;
import java.lang.reflect.Array;
import java.util.Arrays;
import java.util.Scanner;

public class Main
{
    static class Bor
    {
        public class vertex
        {
            int         next[];
            boolean     leaf;
            int         p;
            int         pch;
            int         link;
            int         go[];
            int         bc;
            vertex(){}
        }
        public vertex t[];
        public boolean good[];
        public int k    = 1;
        public int ans  = 0;
        public int bc   = 0;
        public void init(int v)
        {
//        Arrays.fill(t[v].next, -1);
//        Arrays.fill(t[v].go, -1);
            t[v].next   = new int[26];
            t[v].go     = new int[26];
            for(int i = 0; i < 26; i++)
            {
                t[v].next[i]    = -1;
                t[v].go[i]      = -1;
            }
            t[v].p      = -1;
            t[v].leaf   = false;
        }
        public int VertexCount()
        {
            return k;
        }
        public void add_string(final String s)
        {
            int v = 0;
            for(int i = 0; i < s.length(); i++)
            {
                int c = (int) s.charAt(i) - 'a';
                if(t[v].next[c] == -1)
                {
                    init(k);
                    t[v].next[c]    = k;
                    t[k].p          = v;
                    t[k].pch        = c;
                    t[k].bc         = t[v].bc + ((good[c]) ? 0 : 1);
                    ans             += (t[k].bc <= bc) ? 1 : 0;
                    k++;
                }
                v = t[v].next[c];
            }
            t[v].leaf = true;
        }
        public int link(int v)
        {
            if(t[v].link == -1)
            {
                if(v == 0 || t[v].p == 0)
                    t[v].link = 0;
                else t[v].link = go(link(t[v].p), t[v].pch);
            }
            return t[v].link;
        }
        public int go(int v, int c)
        {
            if(t[v].go[c] == -1)
            {
                if(t[v].next[c] != -1)
                    t[v].go[c] = t[v].next[c];
                else if(v == 0)
                    t[v].go[c] = 0;
                else
                    t[v].go[c] = go(link(v), c);
            }
            return t[v].go[c];
        }
        Bor()
        {
            t = new vertex[2000500];
            for(int i = 0; i < t.length; i++)
                t[i] = new vertex();
            good = new boolean[26];
            Arrays.fill(good, false);
            init(0);
        }
    }

    public static void main(String[] argc)
    {
        Scanner sc      = new Scanner(System.in);
        PrintWriter pw  = new PrintWriter(System.out);
        Bor b = new Bor();
        String s1 = sc.next(), s2 = sc.next();
        int k   = sc.nextInt();
        b.bc    = k;
        for(int i = 0; i < 26; i++)
        {
            b.good[i] = (s2.charAt(i) - '0' != 0);
        }
        for(int i = 0; i < s1.length(); i++)
        {
            b.add_string(s1.substring(i));
        }
        pw.println(b.ans);
        sc.close();
        pw.close();
    }
}
